gaps/gap_node_descriptor/gap_node_descriptor.md

```diff
@@ -0,0 +1,109 @@
+
+#### Creating a signature
+
+In order to cryptographically sign a node descriptor the following steps need to be taken:
+
+1. Create a binary representation of the `nodeDescriptor` property by serializing the content via the `JSON Canonicalization Scheme`
+2. Sign the binary data with the chosen signature algorithm using the private key of the signer. The private key must be the one connected to the public key specified in the signer's certificate.
+3. Create the `signature` property in the JSON certificate and add the following details
+    - name of the hash and encryption algorithm (details of the signature algorithm) used to create the signature into the `algorithm` property
+    - binary data of the signature into `value` property
+    - the `signer` property of the `signature` must contain the signing certificate
+
+#### Verifying a signature
+
+1. Verify that the signing certificate contained in `signature.signer` is allowed to sign node descriptors (it's `keyUsage` property is either set to the string `all` or the list contains `signNode`).
+2. Create a binary representation of the `nodeDescriptor` property by serializing the content via the `JSON Canonicalization Scheme`.
+3. Using the signature algorithm matching the one used for creating the signature (`signature.algorithm.hash` and `signature.algorithm.encryption`) and the signer's public key (the `certificate.publicKey` property of the signing certificate), verify that the signature is valid for the binary data obtained in step 2.
+
+### Node descriptor in the Demand
+
+The node descriptor is added as a json object in the `golem.node.descriptor` property of the demand and becomes integral part of the demand during negotiation.
+
+### Verifying during negotiation
+
+This part explains the steps to be taken when verifying a node descriptor during negotiation to validate if it can be used for the demand or not.
+
+1. Verify that the signature on the node descriptor is valid and that the signing [certificate chain is valid](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-25_golem_certificates/gap-25_golem_certificates.md#verifying-a-certificate-chain).
+2. Verify that the moment the negotiation is happening is within the node descriptor's validity period (it is not before the `notBefore` property and not after the `notAfter` property).
+3. Verify that the node descriptor's permission property does not include any permissions that is not granted to the signing certificate. (Details about permission verification can be found in the [Golem certificate structure document](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-25_golem_certificates/gap-25_golem_certificates.md#verifying-a-certificate-chain))
+4. Verify that the Computation manifest does not request any features or access that the node is not granted via the node descriptor.
+
```

Comment:
Also here it must be considered where is the descriptor validation performed, and how can the validation output information be referred to, also by the Demand/Offer language constraint expressions.

This seems an ideal usecase for Property Aspects (https://golem-network.gitbook.io/golem-infrastructure-documentation-develop/architecture/golem-demand-and-offer-specification-language#property-aspects). If `golem.node.descriptor` property contains the descriptor, and that descriptor is validated by the daemon, then the output (as well as any 'signature-specific' data) can be represented (and referred to by constraints) as aspects of that property, eg.:

- `golem.node.descriptor{isValid}` - to represent the validation result
- `golem.node.descriptor{validTo}` - to represent the end of signing certificate validity period,
- etc.

Reply:
I am not sure what you mean as the Demand contains a property `golem.node.descriptor` and that property contains an object, which is the signed node descriptor. It contains all signed properties (including validity, granted features and node Id) with all the certificates required to verify the signatures, validity periods and allowed permissions up to a self signed root.
The implementation can check if there is any certificate in the chain that it trusts so it can accept the node descriptor and process its contents to allow access to restricted features to the node.

The verification cannot happen at a different place/time than the extraction of information from the node descriptor and processing that information, because after the verification the metadata could be hijacked and modified rendering the whole solution unusable.

This is a similar solution to [Payload manifest](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-5_payload_manifest/gap-5_payload_manifest.md) and there are no Property Aspects mentioned.
Also because the node descriptor is added as a JSON object its details can be accessed through normal property paths, like: `golem.node.descriptor.nodeDescriptor.vailidyPeriod` gets the validity period of the node descriptor, I am not sure how aspects would be useful here.
(Same for signing certificate: `golem.node.descriptor.signature.signer.certificate.validityPeriod`.)

Reply:
Right. I'd like us to get into a conversation to cover the ambiguities, specifically such as:
- where in the whole stack is the descriptor validation performed
- how are results of the validation affecting the Demand/Offer matching, eg.
- ...what happens if validation has failed and...
- ...how can results of the validation be referred to by Offer constraints.

@evik42 who do you think should be part of such a walkthrough session? Just yourself, or maybe @nieznanysprawiciel too?

Reply:
@stranger80 we can have a discussion about the reference implementation.

In my opinion this is a GAP which describes a schema for node descriptors and a property to hold these in the demand (define the protocol extension). This GAP should not make assumptions or prescribe anything about the actual implementation.

It is up to the specific implementation to use this property, to decide where verification happens and how to use the data or to completely ignore it.

Could you please provide some example about Offer constraints in the context of Payload manifests? I see payload manifests as a very similar construct to node descriptors, but instead of the node they validate the payload. I did not find any mentions of Offer constraints nor Property Aspects in GAP-5 it would be very helpful if you would provide some example so I better understand what you are aiming for.

I think @nieznanysprawiciel should be there.

Reply:
Invite sent for 9th May.

Reply:
Summary of chat on 9.05.:
- Explained why in current yagna implementation the of the node descriptor validation is easier in provider Agent app, rather than in the yagna daemon.
- The above implies there would be no support for placing node descriptor validity constraints in Demand/Offer (which would require the usage of property aspects and implementing the validation in daemon layer)
- Can we ask for the following enhancements in the GAP article:
  1. include a step-by-step illustrative description (example) of how node descriptor is generated, included in Demand, passing through layers of golem stack, validated on agent side and accepted or discarded
  2. in Rationale section, please make a clear and distinct information about conscious choice not to allow Provider to place constraints on node descriptor validity in the Offer constraints. (Note that it may be added later on)

Reply:
@approxit You can use different certificate type (this is what you name `attestationType` as I understand), by setting different certificate schema:
```
"signer": {
   "$schema": "https://golem.network/schemas/v1/certificate.schema.json",
   ...
}
```